SCREAM: make test-all-scream side-effect-free

bartgol opened the pull request:
This PR makes sure the repo is returned to its original state at the end of test-all-scream execution.

This is only useful if any of the following apply:

- user tests against a baseline ref other than HEAD, and something goes wrong while switching branches, or during the baselines generation
- user does an integration test. If successful, and the branch is not ahead of master, the repo will contain a merge commit. If something goes wrong during the merge, the repo may not be clean.

This mod should not affect a normal development cycle (in particular, it has no effect when running with `--keep-tree` option).

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Bowman to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

jgfouca commented on components/scream/scripts/test_all_scream.py:
```diff
@@ -298,31 +300,41 @@ def run_all_tests(self):
+                if not success:
+                    print ("Error(s) occurred during test phase")
+
+            if not self._keep_tree:
```
This is a perfect usecase for python's `finally` semantics.

```
finally:
    if not self._keep_tree:
        # Cleanup the repo if needed, then re-throw whatever the exception was
        cleanup_repo(self._original_head, self._original_commit)
```

bartgol replied:
Nice. Thanks!

jgfouca commented on components/scream/scripts/test_all_scream.py:
```diff
@@ -298,31 +300,41 @@ def run_all_tests(self):
-            success = self.generate_all_baselines()
-            if not success:
-                print ("Error(s) occurred during baselines generation phase")
-                return success
```
Restore this return, it works with `finally`

bartgol replied:
There's a `return success` at the bottom of the function though, that should be enough, no?

jgfouca replied:
Yes, but the control-flow of this loop is different without the return being here.

bartgol replied:
Ah, right, we want to skip the tests.

jgfouca commented on components/scream/scripts/utils.py:
```diff
@@ -366,7 +366,8 @@ def merge_git_ref(git_ref, repo=None):
     """
-    run_cmd_no_fail("git merge {} -m 'Automatic merge of {}'".format(git_ref,git_ref),from_dir=repo)
+    expect(is_repo_clean(), "Cannot merge ref '{}'. The repo is not clean.".format(git_ref))
+    stat = run_cmd("git merge {} -m 'Automatic merge of {}'".format(git_ref,git_ref),from_dir=repo)
```
This should be run_cmd_no_fail I think. That, or the contents of `stat` should be checked.

bartgol replied:
True. Thanks.

jgfouca commented on components/scream/scripts/utils.py:
```diff
@@ -398,3 +399,33 @@ def checkout_git_ref(git_ref, verbose=False, repo=None):
+    """
+    Get repo toplevel directory
+    """
+    stat, dirname = run_cmd_no_fail("git rev-parse --show-toplevel")
```
run_cmd_no_fail does not return a stat, it only return the output and raises an expect if the command failed.

```
    return run_cmd_no_fail("git rev-parse --show-toplevel"
```

bartgol replied:
True. Thanks.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Blake to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

bartgol commented:
@jgfouca I addressed all your comments except for the `return success` one (see inline comment). I also added an extra commit to improve the processors usage in case of parallel testing.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Blake to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Blake to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Bowman to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

jgfouca commented on components/scream/scripts/test_all_scream.py:
```diff
@@ -298,31 +313,43 @@ def run_all_tests(self):
+                if not success:
+                    print ("Error(s) occurred during test phase")
+
+        except:
```
This is not necessary I don't think. The finally alone should work.

bartgol replied:
makes sense.

bartgol replied:
makes sense.

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_Bowman to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Failure: Timed out waiting for job SCREAM_PullRequest_Autotester_White to start: Total Wait = 1803
- Other jobs have been previously started - We must stop them...

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - User Requested Retest - Label AT: RETEST will be reset after testing.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Testing Jenkins Projects:

<details>
<summary>
  <b>Pull Request Auto Testing STARTING</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 519</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Bowman</h3>
<ul>
  <li>Build Num: 477</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_White</h3>
<ul>
  <li>Build Num: 426</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Using Repos:</h2>

<dl>
  <dt>Repo: SCREAM (E3SM-Project/scream)</dt>
  <ul>
    <li>Branch: testing/minor_mods</li>
    <li>SHA: 7ef30185b69ca15184bcda3508194b9cfabb5036</li>
    <li>Mode: TEST_REPO</li>
  </ul>
</dl>

Pull Request Author: bartgol
</details>

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Jenkins Testing: all Jobs PASSED

<details>
<summary>
  <b>Pull Request Auto Testing has PASSED</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 519</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Bowman</h3>
<ul>
  <li>Build Num: 477</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_White</h3>
<ul>
  <li>Build Num: 426</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>301</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>testing/minor_mods</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>7ef30185b69ca15184bcda3508194b9cfabb5036</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>daebbfbe423a7b7a713005d33fcd2617862a1c08</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>
</details>

E3SM-Autotester commented:
Status Flag 'Pre-Merge Inspection' - SUCCESS: The last commit to this Pull Request has been INSPECTED AND APPROVED by [ jgfouca ]!

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Pull Request will be Automerged

E3SM-Autotester commented:
Merge on Pull Request# 301: IS A SUCCESS - Pull Request successfully merged
